pkg/util/spill/agg_spill_util.go

```diff
@@ -0,0 +1,406 @@
+	retVal := *(*types.MyDecimal)(unsafe.Pointer(&buf[*pos]))
+	*pos += types.MyDecimalStructSize
+	return retVal
+}
+
+// DeserializeTime deserializes Time type
+func DeserializeTime(buf []byte, pos *int64) types.Time {
+	retVal := *(*types.Time)(unsafe.Pointer(&buf[*pos]))
+	*pos += TimeLen
+	return retVal
+}
+
+// DeserializeDuration deserializes Duration type
+func DeserializeDuration(buf []byte, pos *int64) gotime.Duration {
+	retVal := *(*gotime.Duration)(unsafe.Pointer(&buf[*pos]))
+	*pos += DurationLen
+	return retVal
+}
+
+// DeserializeJSONTypeCode deserializes JSONTypeCode type
+func DeserializeJSONTypeCode(buf []byte, pos *int64) types.JSONTypeCode {
+	retVal := *(*types.JSONTypeCode)(unsafe.Pointer(&buf[*pos]))
+	*pos += JSONTypeCodeLen
+	return retVal
+}
+
+// DeserializeInterface deserializes interface type
+func DeserializeInterface(buf []byte, pos *int64) interface{} {
+	// Get type
+	dataType := int(buf[*pos])
+	*pos += InterfaceTypeLen
+
+	switch dataType {
+	case BoolType:
+		res := DeserializeBool(buf, pos)
+		return res
+	case Int64Type:
+		res := DeserializeInt64(buf, pos)
+		return res
+	case Uint64Type:
+		res := DeserializeUint64(buf, pos)
+		return res
+	case FloatType:
+		res := DeserializeFloat64(buf, pos)
+		return res
+	case StringType:
+		strLen := DeserializeInt64(buf, pos)
+		res := string(buf[*pos : *pos+strLen])
+		*pos += strLen
+		return res
+	case BinaryJSONType:
+		retValue := DeserializeBinaryJSON(buf, pos)
+		return retValue
+	case OpaqueType:
+		return DeserializeOpaque(buf, pos)
+	case TimeType:
+		return DeserializeTime(buf, pos)
+	case DurationType:
+		value := DeserializeInt64(buf, pos)
+		fsp := DeserializeInt(buf, pos)
+		return types.Duration{
+			Duration: gotime.Duration(value),
+			Fsp:      fsp,
+		}
+	default:
+		panic("Invalid data type happens in agg spill deserializing!")
+	}
+}
+
+// DeserializeBinaryJSON deserializes Set type and return the size of deserialized object
+func DeserializeBinaryJSON(buf []byte, pos *int64) types.BinaryJSON {
+	retValue := types.BinaryJSON{}
+	retValue.TypeCode = DeserializeJSONTypeCode(buf, pos)
+	jsonValueLen := DeserializeInt(buf, pos)
+	retValue.Value = make([]byte, jsonValueLen)
+	copy(retValue.Value, buf[*pos:*pos+int64(jsonValueLen)])
+	*pos += int64(jsonValueLen)
+	return retValue
+}
+
+// DeserializeSet deserializes Set type
+func DeserializeSet(buf []byte, pos *int64) types.Set {
+	retValue := types.Set{}
+	retValue.Value = DeserializeUint64(buf, pos)
+	nameLen := DeserializeInt(buf, pos)
+	retValue.Name = string(buf[*pos : *pos+int64(nameLen)])
+	*pos += int64(len(retValue.Name))
+	return retValue
+}
+
+// DeserializeEnum deserializes Set type
+func DeserializeEnum(buf []byte, pos *int64) types.Enum {
+	retValue := types.Enum{}
+	retValue.Value = DeserializeUint64(buf, pos)
+	nameLen := DeserializeInt(buf, pos)
+	retValue.Name = string(buf[*pos : *pos+int64(nameLen)])
+	*pos += int64(len(retValue.Name))
```

Comment:
This can be wrapped into DeserializeString
The allocation of `string(buf)` will escape to heap, thus there is no difference whether or not to wrap this code block into a function.
So we can define a `DeserializeString` to make it more readable.

``` go
// File: main.go
// Created Date: 2020-02-01 11:00:00

package main

type fooType struct {
        str string
}

func main(){
}

func DeserializeString2(buf []byte) fooType{
        var s fooType
        var pos *int = new(int)
        s.str = DeserializeString(buf, pos)
        return s
}

func DeserializeString(buf []byte, pos *int) string{
        strLen := 11
        s := string(buf[*pos:*pos+strLen])
        *pos += strLen
        return s
}
```
``` shell
> go build -gcflags '-m -l'
# escape_test
./main.go:20:24: buf does not escape
./main.go:20:36: pos does not escape
./main.go:22:17: string(buf[*pos:*pos + strLen]) escapes to heap
./main.go:13:25: buf does not escape
./main.go:15:20: new(int) does not escape
```

``` go
package main

import (
        "testing"
)

func BenchmarkDeserializeString(b *testing.B) {
        s := []byte("Hello world")
        var pos *int = new(int)
        var a fooType
        for i := 0; i < b.N; i++ {
                *pos = 0
                a.str = DeserializeString(s, pos)
        }
        var _ = a
}

func BenchmarkDeserializeString2(b *testing.B) {
        s := []byte("Hello world")
        var a fooType
        for i := 0; i < b.N; i++ {
                a = DeserializeString2(s)
        }
        var _ = a
}
```

``` shell
> go test -bench=. -benchmem
goos: linux
goarch: amd64
pkg: escape_test
cpu: 13th Gen Intel(R) Core(TM) i7-13700K
BenchmarkDeserializeString-24           100000000               10.97 ns/op           16 B/op          1 allocs/op
BenchmarkDeserializeString2-24          92011054                11.21 ns/op           16 B/op          1 allocs/op
PASS
ok      escape_test     2.155s

```

Reply:
> This can be wrapped into DeserializeString The allocation of `string(buf)` will escape to heap, thus there is no difference whether or not to wrap this code block into a function. So we can define a `DeserializeString` to make it more readable.
> 
> ```go
> // File: main.go
> // Created Date: 2020-02-01 11:00:00
> 
> package main
> 
> type fooType struct {
>         str string
> }
> 
> func main(){
> }
> 
> func DeserializeString2(buf []byte) fooType{
>         var s fooType
>         var pos *int = new(int)
>         s.str = DeserializeString(buf, pos)
>         return s
> }
> 
> func DeserializeString(buf []byte, pos *int) string{
>         strLen := 11
>         s := string(buf[*pos:*pos+strLen])
>         *pos += strLen
>         return s
> }
> ```
> 
> ```shell
> > go build -gcflags '-m -l'
> # escape_test
> ./main.go:20:24: buf does not escape
> ./main.go:20:36: pos does not escape
> ./main.go:22:17: string(buf[*pos:*pos + strLen]) escapes to heap
> ./main.go:13:25: buf does not escape
> ./main.go:15:20: new(int) does not escape
> ```
> 
> ```go
> package main
> 
> import (
>         "testing"
> )
> 
> func BenchmarkDeserializeString(b *testing.B) {
>         s := []byte("Hello world")
>         var pos *int = new(int)
>         var a fooType
>         for i := 0; i < b.N; i++ {
>                 *pos = 0
>                 a.str = DeserializeString(s, pos)
>         }
>         var _ = a
> }
> 
> func BenchmarkDeserializeString2(b *testing.B) {
>         s := []byte("Hello world")
>         var a fooType
>         for i := 0; i < b.N; i++ {
>                 a = DeserializeString2(s)
>         }
>         var _ = a
> }
> ```
> 
> ```shell
> > go test -bench=. -benchmem
> goos: linux
> goarch: amd64
> pkg: escape_test
> cpu: 13th Gen Intel(R) Core(TM) i7-13700K
> BenchmarkDeserializeString-24           100000000               10.97 ns/op           16 B/op          1 allocs/op
> BenchmarkDeserializeString2-24          92011054                11.21 ns/op           16 B/op          1 allocs/op
> PASS
> ok      escape_test     2.155s
> ```

done